Noodle on third-party cookies (#1878)
httpwg/http-extensions: draft-ietf-httpbis-rfc6265bis.md

# Privacy Considerations

The primary privacy risk associated with Cookies is their ability to correlate user activity. This can happen on a single site, but is most problematic when activity is tracked across different, seemingly unconnected Web sites to build a profile of the user.

> Over time, this capability (warned against explicitly in {{?RFC2109}} and all of its successors) has become widely used to authenticate users across sites, to combat fraud and other forms of undesirable traffic, to target advertisements at specific users or at users with specified attributes, to measure how often ads are shown to users, and to recognise when an ad resulted in a change in user behavior.

Reviewer: ```suggestion
Over time, this capability (warned against explicitly in {{?RFC2109}} and all of its successors) has become widely used to correlate user activity across sites, known also as tracking. Reasons are varied and include protection against fraud and other forms of undesirable traffic, assembling information on users, targeting advertisements at specific users or at users with specified attributes, measuring how often ads are shown to users, and to recognize when an ad resulted in a change in user behavior.
```

Author: committing, but it needs some tweaks and you're losing federated auth...

> While in some instances these purposes may not be privacy-invasive, the potential abuses of tracking imply a need for significant limitations on cookies. However, any such limitations need to be imposed with care, because cookies are widely used for a variety of purposes that are critical to the functioning of a Web site (and the Web).

Reviewer: I think that you are making the wrong distinction.  The purposes are always tracking, but you are implying a value judgment they are objectionable from a privacy standpoint (though this is of course subjective).

```suggestion
Though some of these purposes might not be objectionable from a privacy perspective, the potential for this tracking capability of cookies to be abused is being carefully reviewed. Some client implementations are deploying constraints on cookies with an aim to limit the potential for abuse.  These limitations need to be carefully balanced against the disruption that might result, especially where Web sites depend on preexisting features of cookies for critical functions.
```

Author: Went in a different direction - see revision.

> As a result, this document recommends limited mitigations against the privacy risks associated with cookies below. It is expected that future updates will impose stricter and more well-defined limitations on cookies, based upon experiments now being undertaken in user agents. Note that some functions currently relying on cookies might eventually be supported by separate, targeted mechanisms.

Reviewer: ```suggestion
This document describes limited mitigations against the privacy risks associated with cookies below. These mitigations are the most widely implemented and compatible at the time of writing.  It is expected that implementations will impose stricter and more well-defined limitations on cookies.  Future versions of this document might codify those limitations based upon experience with those limitations. If functions that currently rely on cookies can be supported by separate, targeted mechanisms, stricter limitations might be more feasible.
```

Cookies are not the only mechanism that can be used to track users across sites, so while these mitigations are necessary to improve Web privacy, they are not sufficient on their own.

## Third-Party Cookies {#third-party-cookies}

A cookie that is sent to a site other than the one it originated from is known as a "third-party"

> cookie. When rendering a document, a user agent can request resources from sites other than that

Reviewer: This first sentence is not right, though the remainder of the definition is accurate.  Cookies are always sent back to the entity that set them, but a third-party cookie is a cookie set by a server that is other than the primary entity with which a user interacts.

Maybe: 

> In a Web browsing context, a "third-party" or cross-site cookie is a cookie that is associated with content that is served from a different server than the one that hosts the primary resource (that is, the HTML).  HTML documents can embed content (such as scripts, images, stylesheets, frames) that is served from other domains.  These "third-party" resources [...]

By the way, [Tess recommends "cross-site"](https://tess.oconnor.cx/2020/10/parties).

Author: Thanks, I knew terminology was going to be an issue...

where that document was obtained from (such as advertising networks). Those third-party resources

> can use cookies to track the user even if they never visit that server directly. For example, if a

Reviewer: The key here is that third-party resources will get the same cookies no matter how they are embedded.  That's the key piece, and this doesn't quite get to that.  If the third-party server sees the same cookie in a variety of contexts, they can link those contexts to the same person (and thereby track them).

Author: I see how that's central to Mozilla's approach, but I don't think this level of detail / specificity is necessary here.

user visits a site that contains content from a third party and then later visits another site that
contains content from the same third party, the third party can track the user between the two
sites.

Because of their inherent privacy issues, most user agents now limit third-party cookies in a
variety of ways. Some completely block third-party cookies by refusing to process third-party
Set-Cookie header fields and refusing to send third-party Cookie header fields. Some partition
cookies based upon the first-party context, so that different cookies are sent depending on the
site being browsed. Some block cookies based upon user agent cookie policy and/or user controls.

It is RECOMMENDED that user agents implement third-party cookie policies that balance the
privacy and compatibility needs of their users. However, this document does not yet endorse any
particular third-party cookie policy.

> As a result, resources SHOULD NOT rely upon third-party cookies functioning reliably.

Reviewer: Isn't this purely factual?  That is, s/SHOULD NOT/cannot/ ?

A normative statement implies that this is the way we want it to be.  I don't think that would be right.

Author: People assign somewhat magical qualities to 2119 language; I'm invoking that here.

Reviewer: I'm saying that I don't think that invoking the mystical 2119 invocation is right.  You are saying that they are unreliable, which is a fact.  But by invoking 2119 you are saying that people shouldn't use third-party cookies, which I don't agree with.  Though perhaps the fine folks at Apple would disagree, I would argue that partitioned cookies are perfectly fine (partitioned storage more generally).  Useful even.

That's an ongoing debate that you would have this draft take a position on.

Author: Fair enough.

Note that cookies are not the only mechanism that can be used to track users across sites, so while these mitigations are necessary to improve Web privacy, they are not sufficient on their own.

## Third-Party Cookies {#third-party-cookies}

In a Web browsing context, a "third-party" or cross-site cookie is a cookie that is associated with
content that is served from a different server than the one that hosts the primary resource (that
is, the HTML). HTML documents can embed content (such as scripts, images, stylesheets, frames) that
is served from other domains. These third-party resources can use cookies to track the user even if
they never visit that server directly. For example, if a user visits a site that contains content
from a third party and then later visits another site that contains content from the same third

> party, the third party can track the user between the two sites.

Reviewer: ```suggestion
party, that third party might be able to use cookies to track the user between the two sites.
```

Author: what value does 'might be able to' add?

Reviewer: ...because "can" implies an absolute possibility (I can use legs to climb stairs), in this case whether tracking is possible depends on what policies are in place with respect to cross site cookies.

Author: This para is already super-wordy. Will try to refactor.

Because of their inherent privacy issues, most user agents now limit third-party cookies in a
variety of ways. Some completely block third-party cookies by refusing to process third-party
Set-Cookie header fields and refusing to send third-party Cookie header fields. Some partition
cookies based upon the first-party context, so that different cookies are sent depending on the
site being browsed. Some block cookies based upon user agent cookie policy and/or user controls.

> It is RECOMMENDED that user agents implement third-party cookie policies that balance the
> privacy and compatibility needs of their users. However, this document does not yet endorse any
> particular third-party cookie policy.

Reviewer: This is a bit of a non-statement really.  I'm all for magical thinking, but wouldn't this be better removed?  You already explain that things are unreliable.

If you want to recommend something, how about:

```suggestion
As the situation is fluid, this document does not endorse any particular policy for third-party cookies.  However, user agents SHOULD adopt a policy for third-party cookies that is as restrictive as compatibility constraints permit, thereby providing the best possible privacy.  Any such policy might become stricter as compatibility constraints change over time.

Consequently, resources cannot rely on a specific treatment for cookies in third-party contexts.
```

Author: I see it more as air cover than magical thinking. See edits.

# Privacy Considerations

> The primary privacy risk associated with cookies is their ability to correlate user activity. This

Author: ```suggestion
Cookies' primary privacy risk is their ability to correlate user activity. This
```

can happen on a single site, but is most problematic when activity is tracked across different,

> seemingly unconnected Web sites to build a profile of the user.

Author: ```suggestion
seemingly unconnected Web sites to build a user profile.
```

Over time, this capability (warned against explicitly in {{?RFC2109}} and all of its successors)
has become widely used for varied reasons including:

* authenticating users across sites,
* assembling information on users,
* protecting against fraud and other forms of undesirable traffic,
* targeting advertisements at specific users or at users with specified attributes,
* measuring how often ads are shown to users, and
* recognizing when an ad resulted in a change in user behavior.

While not every use of cookies is necessarily problematic for privacy, their potential for abuse

> has become a widespread concern in industry and society. In response to these concerns, user agents

Author: ```suggestion
has become a widespread concern in the Internet community and broader society. In response to these concerns, user agents
```

I think this is more appropriate -- the IETF isn't just industry

have actively constrained cookie functionality in various ways (as allowed and encouraged by
previous specifications), while avoiding disruption to features they judge desirable for the health
of the Web.

> It is too early to declare consensus on which specific mechanism(s) should be used to mitigate the privacy impact of cookies; ongoing developments in user agents are best characterised as experiments that

Reviewer 2: I think some browsers are well beyond what would be characterized as experiments.

> It is too early to declare consensus on which specific mechanism(s) should be used to mitigate the privacy impact of cookies; ongoing developments in user agents are best characterised as experiments that

Author: ```suggestion
It is too early to declare consensus on which specific mechanism(s) should be used to mitigate cookies' privacy impact; user agents' ongoing changes in how they are handled are best characterised as experiments that
```

can provide input into that eventual consensus.

Instead, this document describes limited, general mitigations against the privacy risks associated
with cookies that enjoy wide deployment at the time of writing. It is expected that implementations
will continue to experiment and impose stricter, more well-defined limitations on cookies over
time. Future versions of this document might codify those mechanisms based upon deployment
experience. If functions that currently rely on cookies can be supported by separate, targeted
mechanisms, they might be documented in separate specifications and stricter limitations on cookies
might become feasible.

Note that cookies are not the only mechanism that can be used to track users across sites, so while
these mitigations are necessary to improve Web privacy, they are not sufficient on their own.

## Third-Party Cookies {#third-party-cookies}

> A "third-party" or cross-site cookie is a cookie that is associated with embedded content (such as

Author: ```suggestion
A "third-party" or cross-site cookie is one that is associated with embedded content (such as
```

# Privacy Considerations

The primary privacy risk associated with cookies is their ability to correlate user activity. This
can happen on a single site, but is most problematic when activity is tracked across different,
seemingly unconnected Web sites to build a profile of the user.

Over time, this capability (warned against explicitly in {{?RFC2109}} and all of its successors)
has become widely used for varied reasons including:

* authenticating users across sites,
* assembling information on users,
* protecting against fraud and other forms of undesirable traffic,
* targeting advertisements at specific users or at users with specified attributes,
* measuring how often ads are shown to users, and
* recognizing when an ad resulted in a change in user behavior.

While not every use of cookies is necessarily problematic for privacy, their potential for abuse
has become a widespread concern in industry and society. In response to these concerns, user agents
have actively constrained cookie functionality in various ways (as allowed and encouraged by
previous specifications), while avoiding disruption to features they judge desirable for the health
of the Web.

> It is too early to declare consensus on how the privacy impact of cookies can be successfully
> mitigated in detail; ongoing developments in user agents are best characterised as experiments that

Author: Slight clarification based on private feedback;

```suggestion
It is too early to declare consensus on which specific mechanism(s) should be used to mitigate the privacy impact of cookies; ongoing developments in user agents are best characterised as experiments that
```
